File: day15/day15.cpp
#include <algorithm>
#include <fstream>
#include <functional>
#include <queue>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Map
{
	vector<int> data;
	size_t width;
	size_t height;

	size_t dijkstra(size_t sx, size_t sy) const
	{
		static const pair<int,int> NEIGH[] = { {-1, 0}, {0, -1}, {1, 0}, {0, 1}};

		vector<int> riskmap(data.size(), 0);
		typedef tuple<int, size_t, size_t> ht;
		priority_queue<ht, vector<ht>, greater<ht>> queue;

		queue.push(make_tuple(0, sx, sy));
		while (!queue.empty())
		{
			auto [risk, x, y] = queue.top();
			queue.pop();

			for (auto [dx, dy]: NEIGH)
			{
				size_t nx = x + dx;
				size_t ny = y + dy;
				if (nx < width && ny < height)
				{
					size_t off = ny * width + nx;
					int nrisk = risk + data[off];
					if (!riskmap[off] || nrisk < riskmap[off])
					{
						riskmap[off] = nrisk;
						queue.push(make_tuple(nrisk, nx, ny));
					}
				}
			}
		}

		return riskmap.back();
	}

	void extend(int n)
	{
		vector<int> ndata(width * height * n * n, 0);
		auto it = ndata.begin();
		for (int ry = 0; ry < n; ry++)
		{
			for (size_t y = 0; y < height; y++)
			{
				for (int rx = 0; rx < n; rx++)
				{
					for (size_t x = 0; x < width; x++)
					{
						*it = (data[y * width + x] + rx + ry - 1) % 9 + 1;
						++it;
					}
				}
			}
		}
		width *= n;
		height *= n;
		data = ndata;
	}
};

istream& operator>>(istream& in, Map& m)
{
	m.data.clear();
	m.width = m.height = 0;

	char ch;
	while (in.get(ch))
	{
		if ('1' <= ch && ch <= '9')
		{
			if (!m.height) { m.width++; };
			m.data.push_back(ch - '0');
		}
		else if (ch == '\n')
		{
			m.height++;
		}
	}
	return in;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Map m;
	in>>m;
	in.close();

	fmt::print("Part1: {}\n", m.dijkstra(0, 0));
	m.extend(5);
	fmt::print("Part2: {}\n", m.dijkstra(0, 0));

	return 0;
}

File: day15/day15.py
#!/usr/bin/env python3

import heapq

class Map(object):
    NEIGH = ((-1,0), (0,-1), (1,0), (0, 1))

    def __init__(self, txt):
        self.m = [[int(x) for x in row] for row in txt.splitlines()]
        self.height = len(self.m)
        self.width = len(self.m[0])

    def extend(self, n):
        m = []
        for y in range(self.height * n):
            row = []
            ry = y % self.height
            qy = y // self.height
            for x in range(self.width * n):
                rx = x % self.width
                qx = x // self.width

                row.append((self.m[ry][rx] + qx + qy - 1) % 9 + 1)

            m.append(row)

        self.m = m
        self.height *= n
        self.width *= n

    def dijkstra(self, start):
        dist = [[0 for x in row] for row in self.m]
        queue = []
        heapq.heappush(queue, (0, start))
        while queue:
            risk, pos = heapq.heappop(queue)
            for dx, dy in self.NEIGH:
                nx, ny = npos = pos[0] + dx, pos[1] + dy
                if 0 <= nx < self.width and 0 <= ny < self.height:
                    nrisk = risk + self.m[ny][nx]
                    if not dist[ny][nx] or nrisk < dist[ny][nx]:
                        dist[ny][nx] = nrisk
                        heapq.heappush(queue, (nrisk, npos))

        return dist[-1][-1]

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        m = Map(txt)
        print("Part1:", m.dijkstra((0, 0)))
        m.extend(5)
        print("Part2:", m.dijkstra((0, 0)))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
